stratum/docs/setup_guide_barefoot_tofino_onos.md

```diff
@@ -0,0 +1,210 @@
+
+The coarse steps are as follows:
+- On host: build ONL
+- On switch: install built ONL
+- On switch: start Stratum
+- On host: start ONOS
+- On host: build `fabric-tofino` ONOS app
+- On host: install and activate `fabric-tofino` app in ONOS
+- On host: register switch in ONOS using `netcfg`
+
+
+## Build or download ONL
+
+You have two choices to obtain an ONL image.
+
+### 1) Building ONL yourself
+
+ONL can be built wherever you like.
+
+```
+git clone --depth 1  --single-branch --branch onf-ONLPv2 https://github.com/opennetworkinglab/OpenNetworkLinux
+cd OpenNetworkLinux
+VERSION=9 make docker
+# or
+sudo VERSION=9 make docker
+
+find -type f -name "*INSTALLER"
+```
+There should be an installer file with a filename similar to:
+
+`./RELEASE/stretch/amd64/ONL-ONLPv2_ONL-OS_2020-06-22.1054-bb20e13_AMD64_INSTALLED_INSTALLER`.
+
+This is the ONL image you have to install on the switch in the next step. 
+
+### 2) Using a pre-built ONL
+
+You can also download a pre-built ONL from [here](https://github.com/opennetworkinglab/OpenNetworkLinux/releases).
+
+:warning: You have to build or use the ONL from the fork mentioned above - if you use a pre-built installer by ONL itself, you will run into trouble with Stratum.
+
+
+## Install ONL using ONIE
+
+- Restart switch with a serial console cable attached
+- Wait for the GNU GRUB selection screen on boot-up
+- Select `ONIE`
+- Select `ONIE: rescue`
+- Install the ONL image created in the previous step by either
+    - exposing the `INSTALLER` via HTTP
+        - `onie-nos-install http://the-http-server-where-the-installer-is-hosted/ONL-ONLPv2_ONL-OS_2020-06-22.1054-bb20e13_AMD64_INSTALLED_INSTALLER`
+    - copying the `INSTALLER` to the switch using scp
+        - `scp YOUR_HOST:ONL-ONLPv2_ONL-OS_2020-06-22.1054-bb20e13_AMD64_INSTALLED_INSTALLER .`
+        - `onie-nos-install ONL-ONLPv2_ONL-OS_2020-06-22.1054-bb20e13_AMD64_INSTALLED_INSTALLER`
+
+The switch should restart after ONIE finished installing ONL.
+
+:pencil2: The `user:password` for ONL is `root:onl`.
+
+## Start Stratum with Docker
+
+On the switch, install [Docker](https://linuxize.com/post/how-to-install-and-use-docker-on-debian-9/).
+Docker is already installed in the pre-built version from the fork mentioned above.
+
+```
+git clone git@github.com:stratum/stratum.git # or checkout a specific branch/tag via --branch some_branch
+cd stratum
+./stratum/hal/bin/barefoot/docker/start-stratum-container.sh
+```
+
+This will start Stratum in the foreground (not detached).
+
+:warning: **Important**: Huge table allocation **must** be activated on the host system, see [here](https://github.com/stratum/stratum/blob/master/stratum/hal/bin/barefoot/README.md#huge-pages--dma-allocation-error).
+
+:pencil2: You can tunnel the ports exposed by Stratum (P4Runtime, gNMI, gNOI) via SSH tunnels:
```

Comment:
I think it would be nice to explain why we need to access these ports via a tunnel (e.g., firewall, ...) and we need to let users know this step is optional.

Reply:
The explanation is provided in the next paragraph:
> This tunnels the ports to your local machine which enables you to execute ONOS commands locally

What would be a more appropriate explanation? 

I will add a notice that tunneling the ports is optional